### PR TITLE
Check if the user ran compile and otherwise run it for them

### LINKED ISSUE #5832: changed application.properties not taken into account starting devmode
`mvn quarkus:devmode` picks up .java changes but does not pick up changes to `application.properties` for that I need to explicitly do `mvn compile quarkus:devmode`.

I would expect it to be picked up if .java changes are OR that all "compilable" source changes will be ignored by `quarkus:devmode`

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Fixes #5832

### CODE REVIEW

**Review by gastaldi — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
Maybe we should also change the docs to leave off compile?

**Comment by gastaldi** [comment]
Yeah, that would be a good thing. Although I haven't been able to reproduce the error Max is seeing even without this PR

**Comment by stuartwdouglas** [comment]
I think the steps are:

- new project
- add application.properties, make change
- mvn quarkus:dev, compile is run because first build, changes are picked up
- stop
- make changes to application.properties
- mvn quarkus:dev, change is not picked up, as compile has not copied the resource

File: devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java [hunk 1]
```diff
@@ -225,8 +249,20 @@ public void execute() throws MojoFailureException, MojoExecutionException {
         if (!sourceDir.isDirectory()) {
             getLog().warn("The project's sources directory does not exist " + sourceDir);
         }
+        //we check to see if there was a compile (or later) goal before this plugin
```
Comment (gastaldi) [inline devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java]: I think you only need to check if `compile` was executed? Not sure if you can `install` or `deploy` without the `compile` phase is invoked

**Comment by gastaldi** [comment]
> I think the steps are:
> 
> * new project
> * add application.properties, make change
> * mvn quarkus:dev, compile is run because first build, changes are picked up
> * stop
> * make changes to application.properties
> * mvn quarkus:dev, change is not picked up, as compile has not copied the resource

I followed your exact steps and still works for me.

**Comment by gastaldi** [comment]
PS: I am using VI, so it's unlikely that an IDE copied the file to the classes folder

File: devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java (hunk 1 above)
Comment (gastaldi) [reply devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java]: Just realized these are the goals set from the `mvn` command. Looks ok

**Comment by machi1990** [comment]
> I think the steps are:
> 
> * new project
> * add application.properties, make change
> * mvn quarkus:dev, compile is run because first build, changes are picked up
> * stop
> * make changes to application.properties
> * mvn quarkus:dev, change is not picked up, as compile has not copied the resource

I guess this will work either way since we have:

https://github.com/quarkusio/quarkus/blob/d9bc53056c3647e627a2f21ea74a5355e3b1ebd1/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java#L231-L235 

@evanchooly added the auto-compile IIRC.  

Maybe we just need to change the docs and drop `compile` ?

**Comment by stuartwdouglas** [comment]
The current code only auto compiles if target/classes does not exist. If they exist but are out of date  there will not be a compile.

**Review by gsmet — APPROVED** [review]
This one looks good to go IMHO.

I rebased it on top of master as it was a bit old. Let's wait for CI.

**Pull request merged**